### PR TITLE
Migrate storage

### PR DESCRIPTION
**Pull request by wolverineks**

### CODE REVIEW

File: src/HW/LedgerConnect/LedgerConnect.tsx [hunk 1]
```diff
@@ -284,7 +284,10 @@ class _LedgerConnect extends React.Component<Props, State> {
         </View>
         {useUSB === true && (
           <Button
-            onPress={() => this._onConfirm(deviceObj)}
+            onPress={() => {
+              if (!deviceObj) throw new Error('invalid state')
```
Comment (wolverineks) [inline src/HW/LedgerConnect/LedgerConnect.tsx]: typescript should have caught this?

File: src/yoroi-wallets/cardano/ShelleyWallet.test.ts [hunk 2]
```diff
@@ -55,7 +52,7 @@ describe('migration', () => {
           }
           `)
 
-    expect(getBech32InternalChain(wallet)).resolves.toMatchInlineSnapshot(
+    await expect(getBech32InternalChain(wallet)).resolves.toMatchInlineSnapshot(
```
Comment (wolverineks) [inline src/yoroi-wallets/cardano/ShelleyWallet.test.ts:56]: async expectations should use `await`

File: src/yoroi-wallets/cardano/shelley/transactionCache.test.ts [hunk 3]
```diff
@@ -74,22 +74,46 @@ describe('transactionCache', () => {
...
+    if (path === 'txids') return [mockTx.id]
     throw new Error('invalid path')
   },
   multiGet: async (txids: Array<string>) => {
-    if (txids.length > 1 || txids[0] !== mockTx.id) throw new Error('invalid path')
+    if (txids[0] !== mockTx.id) throw new Error('invalid path')
 
     return [
-      [txids[0], JSON.stringify(mockTx)], //
+      [txids[0], mockTx] as [string, Transaction], //
```
Comment (wolverineks) [inline src/yoroi-wallets/cardano/shelley/transactionCache.test.ts:108]: storage auto stringifies/parses

File: src/yoroi-wallets/cardano/shelley/transactionCache.ts [hunk 4]
```diff
@@ -497,13 +491,12 @@ const makeTxCacheStorage = (storage: Storage): TxCacheStorage => ({
   },
 
   saveTxs: async (txs: TransactionCacheState['transactions']) => {
-    const items: Array<[txid: string, transaction: string]> = Object.entries(txs) //
-      .map(([txid, tx]) => [txid, JSON.stringify(tx)])
+    const items = Object.entries(txs)
```
Comment (wolverineks) [inline src/yoroi-wallets/cardano/shelley/transactionCache.ts:494]: storage auto stringifies

File: src/yoroi-wallets/cardano/shelley/transactionCache.ts [hunk 5]
```diff
@@ -497,13 +491,12 @@ const makeTxCacheStorage = (storage: Storage): TxCacheStorage => ({
...
     const txids = Object.keys(txs)
 
     await Promise.all([
       storage.multiSet(items), //
-      storage.setItem('txids', JSON.stringify(txids)),
+      storage.setItem('txids', txids),
```
Comment (wolverineks) [inline src/yoroi-wallets/cardano/shelley/transactionCache.ts:499]: storage auto stringifies

File: src/yoroi-wallets/cardano/ShelleyWallet.test.ts [hunk 6]
```diff
@@ -1,33 +1,31 @@
+/* eslint-disable @typescript-eslint/no-explicit-any */
```
Comment (wolverineks) [inline src/yoroi-wallets/cardano/ShelleyWallet.test.ts:1]: enable testing private properties

File: src/yoroi-wallets/cardano/ShelleyWallet.test.ts [hunk 7]
```diff
@@ -539,9 +525,7 @@ const data: WalletJSON = {
   provider: '',
 }
 
-const getBech32InternalChain = (wallet: ShelleyWalletLegacy) =>
-  wallet.internalChain?._addressGenerator._accountPubKeyPtr?.toBech32()
-const getRewardAddress = (wallet: ShelleyWalletLegacy) =>
-  wallet.getRewardAddress().then((address) => address.toBech32())
-const getWalletData = (wallet: ShelleyWalletLegacy) => wallet.storage.read(`/wallet/${wallet.id}/data`)
-const getStakingKey = (wallet: ShelleyWalletLegacy) => wallet.getStakingKey().then((key) => key.toBech32())
+const getBech32InternalChain = (wallet: any) => wallet.internalChain?._addressGenerator._accountPubKeyPtr?.toBech32()
```
Comment (wolverineks) [inline src/yoroi-wallets/cardano/ShelleyWallet.test.ts:528]: enable testing private properties

File: src/Storage/migrations/walletMeta.ts [hunk 8]
```diff
@@ -105,7 +106,13 @@ function migrateAttribute(attr: keyof WalletMeta, walletMetaUpdated: Partial<Wal
...
 }
 
-function isWalletMeta(walletMeta: WalletMeta | object | undefined): walletMeta is WalletMeta {
+export const parseWalletMeta = (data: unknown) => {
+  const parsed = parseSafe(data)
+
+  return isWalletMeta(parsed) ? parsed : undefined
+}
+
+export function isWalletMeta(walletMeta: any): walletMeta is WalletMeta {
```
Comment (stackchain) [inline src/Storage/migrations/walletMeta.ts:115]:
```suggestion
export function isWalletMeta(walletMeta: unknown): walletMeta is WalletMeta {
```

File: src/yoroi-wallets/cardano/ShelleyWallet.ts [hunk 9]
```diff
@@ -112,7 +112,7 @@ export type WalletJSON = ShelleyWalletJSON | ByronWalletJSON
 
 export default ShelleyWallet
 export class ShelleyWallet implements WalletInterface {
-  storage: typeof storageLegacy
+  storage: Storage
```
Comment (stackchain) [inline src/yoroi-wallets/cardano/ShelleyWallet.ts]:
```suggestion
  readonly storage: Storage
```

File: src/yoroi-wallets/cardano/utxoManager.test.ts [hunk 10]
```diff
@@ -0,0 +1,106 @@
+import {storage as rootStorage} from '../storage'
+import {makeUtxoManagerStorage, makeUtxoStorage} from './utxoManager'
+
+describe('utxo manager storage', () => {
+  it('works', async () => {
+    const storage = rootStorage.join('utxos/')
+    const {addrCounter} = makeUtxoManagerStorage(storage)
+
+    // iniital
```
Comment (stackchain) [inline src/yoroi-wallets/cardano/utxoManager.test.ts]:
```suggestion
    // initial
```

File: src/yoroi-wallets/cardano/utxoManager.ts [hunk 11]
```diff
@@ -138,3 +63,55 @@ const serializer = (utxo: UtxoModels.Utxo): RawUtxo => ({
...
+const diffPath = 'diff'
+const safePointPath = 'safe-point'
+
+export const makeUtxoStorage = (storage: Storage) => {
+  const getUtxoDiffToBestBlock = () => storage.getItem(diffPath, parseDiff).then((diff) => diff ?? [])
+  const setUtxoDiffToBestBlock = async (utxoDiffToBestBlock: UtxoDiffToBestBlock[]) =>
+    storage.setItem(diffPath, utxoDiffToBestBlock)
+
+  const getUtxoAtSafePoint = () => storage.getItem(safePointPath, parseSafePoint)
+  const setUtxoAtSafePoint = async (utxoAtSafePoint: UtxoAtSafePoint) => storage.setItem(safePointPath, utxoAtSafePoint)
```
Comment (stackchain) [inline src/yoroi-wallets/cardano/utxoManager.ts]:
```suggestion
  const setUtxoAtSafePoint = (utxoAtSafePoint: UtxoAtSafePoint) => storage.setItem(safePointPath, utxoAtSafePoint)
```

File: src/yoroi-wallets/cardano/utxoManager.ts [hunk 12]
```diff
@@ -138,3 +63,55 @@ const serializer = (utxo: UtxoModels.Utxo): RawUtxo => ({
...
+
+  const utxoStorage: UtxoStorage = {
+    getUtxoAtSafePoint,
+    replaceUtxoAtSafePoint: async (utxos: Utxo[], lastSafeBlockHash: string) =>
```
Comment (stackchain) [inline src/yoroi-wallets/cardano/utxoManager.ts]:
```suggestion
    replaceUtxoAtSafePoint: (utxos: Utxo[], lastSafeBlockHash: string) =>
```

File: src/yoroi-wallets/cardano/utxoManager.ts [hunk 13]
```diff
@@ -138,3 +63,55 @@ const serializer = (utxo: UtxoModels.Utxo): RawUtxo => ({
 })
 
 export type UtxoManager = Awaited<ReturnType<typeof makeUtxoManager>>
+
+const diffPath = 'diff'
+const safePointPath = 'safe-point'
+
+export const makeUtxoStorage = (storage: Storage) => {
+  const getUtxoDiffToBestBlock = () => storage.getItem(diffPath, parseDiff).then((diff) => diff ?? [])
+  const setUtxoDiffToBestBlock = async (utxoDiffToBestBlock: UtxoDiffToBestBlock[]) =>
```
Comment (stackchain) [inline src/yoroi-wallets/cardano/utxoManager.ts]:
```suggestion
  const setUtxoDiffToBestBlock = (utxoDiffToBestBlock: UtxoDiffToBestBlock[]) =>
```

File: src/yoroi-wallets/storage.ts [hunk 14]
```diff
@@ -1,46 +1,62 @@
...
+  }
 
   return {
-    getItem: (key: string) => {
-      return AsyncStorage.getItem(withPrefix(key))
-    },
-    multiGet: async (keys: Array<string>) => {
-      const prefixedKeys = keys.map((key) => withPrefix(key))
-      const items = await AsyncStorage.multiGet(prefixedKeys)
+    join: (folderName: Path) => mountStorage(`${path}${folderName}`),
```
Comment (stackchain) [inline src/yoroi-wallets/storage.ts:28]: I'm guessing that the `Path` type, dismisses the necessity to test paths ending with`/`.

File: src/yoroi-wallets/walletManager.ts [hunk 15]
```diff
@@ -34,19 +34,21 @@ export class WalletManager {
...
+    this.storage = storage.join('wallet/')
   }
 
   async listWallets() {
-    const keys = await storage.keys('/wallet/')
-    const result = await Promise.all(keys.map((key) => storage.read<WalletMeta>(`/wallet/${key}`)))
+    const keys = await this.storage.getAllKeys()
+    const result = await this.storage.multiGet(keys, parseWalletMeta)
 
     Logger.debug('result::_listWallets', result)
```
Comment (stackchain) [inline src/yoroi-wallets/walletManager.ts]: maybe this doesn't make sense anymore, in case some wallet is corrupted.

Comment (stackchain) [inline src/yoroi-wallets/walletManager.ts]:
```suggestion
```

File: src/yoroi-wallets/walletManager.ts [hunk 16]
```diff
@@ -252,18 +258,18 @@ export class WalletManager {
...
     // wallet.remove
-    await storage.remove(`/wallet/${id}/data`)
-    await EncryptedStorage.remove(EncryptedStorageKeys.rootKey(id))
-    await Keychain.removeWalletKey(id)
 
-    await storage.remove(`/wallet/${id}`)
+    await this.storage.removeItem(`${id}/data`) // remove wallet data
+    await EncryptedStorage.remove(EncryptedStorageKeys.rootKey(id)) // remove auth with password
+    await Keychain.removeWalletKey(id) // remove auth with os
+    await this.storage.removeItem(id) // remove wallet meta
```
Comment (stackchain) [inline src/yoroi-wallets/walletManager.ts:266]: can we have some tests for the walletManager, at least for create/remove?

**Review by stackchain — APPROVED** [review]

File: src/yoroi-wallets/cardano/ShelleyWallet.ts [hunk 17]
```diff
@@ -112,25 +112,26 @@ export type WalletJSON = ShelleyWalletJSON | ByronWalletJSON
...
-  protected encryptedStorage: WalletEncryptedStorage
-  readonly primaryToken: Readonly<DefaultAsset>
-  id: string
-  networkId: NetworkId
-  walletImplementationId: WalletImplementationId
-  isHW = false
-  hwDeviceInfo: null | HWDeviceInfo
-  isReadOnly: boolean
-  provider: null | undefined | YoroiProvider
+  readonly primaryToken: DefaultAsset
```
Comment (stackchain) [inline src/yoroi-wallets/cardano/ShelleyWallet.ts:115]:
```suggestion
  readonly primaryToken: Readonly<DefaultAsset>
```

File: src/yoroi-wallets/cardano/ShelleyWallet.ts [hunk 18]
```diff
@@ -112,25 +112,26 @@ export type WalletJSON = ShelleyWalletJSON | ByronWalletJSON
...
   isEasyConfirmationEnabled = false
-  internalChain: AddressChain
-  externalChain: AddressChain
-  publicKeyHex: string
-  rewardAddressHex: null | string = null
-  version: string
-  checksum: CardanoTypes.WalletChecksum
+
   private _utxos: RawUtxo[]
+  private readonly storage: Storage
```
Comment (stackchain) [inline src/yoroi-wallets/cardano/ShelleyWallet.ts:132]:
```suggestion
  private readonly storage: Readonly<Storage>
```

**Review by banklesss — APPROVED** [review]

File: src/yoroi-wallets/cardano/utxoManager.ts [hunk 19]
```diff
@@ -138,3 +63,55 @@ const serializer = (utxo: UtxoModels.Utxo): RawUtxo => ({
...
+  return utxoStorage
+}
+
+const parseSafePoint = (data: unknown) => {
+  const parsed = parseSafe(data)
+  return isSafePoint(parsed) ? parsed : undefined
+}
+const isSafePoint = (data: unknown): data is UtxoAtSafePoint => {
+  const candidate = data as UtxoAtSafePoint
+  return !!candidate && !!candidate.lastSafeBlockHash && Array.isArray(candidate.utxos)
```
Comment (michalszorad) [inline src/yoroi-wallets/cardano/utxoManager.ts:107]: The type of `lastSafeBlockHash` or the type of items in `utxos` is not checked. This is an issue we face in multiple places of the app. Can we use some parsing library going forward? I've built this https://www.npmjs.com/package/create-typeguard couple of years ago, but I'm sure there are better alternatives out there

File: src/yoroi-wallets/storage.test.ts [hunk 20]
```diff
@@ -0,0 +1,118 @@
...
+
+describe('prefixed storage', () => {
+  beforeEach(() => AsyncStorage.clear())
+
+  it('getAllKeys, setItem, getItem, removeItem, clear', async () => {
+    const storage = yoroiStorage
+    await storage.getAllKeys().then((keys) => expect(keys).toEqual([]))
+
+    await storage.setItem('item1', item1)
+    await storage.getItem('item1').then((item) => expect(item).toEqual(item1))
```
Comment (michalszorad) [inline src/yoroi-wallets/storage.test.ts]: isn't `expect( await storage.getItem('item1')).toEqual(item1)` more readable?

File: storybook/mocks/wallet.ts [hunk 21]
```diff
@@ -36,7 +37,7 @@ const walletMeta: WalletMeta = {
   walletImplementationId: 'haskell-shelley-24',
 }
 
-const wallet: YoroiWallet = {
+const wallet: YoroiWallet & Record<string, unknown> = {
```
Comment (michalszorad) [inline storybook/mocks/wallet.ts]: Why is the `& Record<string, unknown>` needed here?

File: src/yoroi-wallets/storage.test.ts [hunk 22]
```diff
@@ -0,0 +1,118 @@
...
+
+    await storage.removeItem('item1')
+    await storage.getItem('item1').then((item) => expect(item).toEqual(null))
+    await storage.getAllKeys().then((keys) => expect(keys).toEqual(['item2']))
+
+    await storage.clear()
+    await storage.getAllKeys().then((keys) => expect(keys).toEqual([]))
+  })
+
+  it('getAllKeys, setItem, getItem, removeItem, clear', async () => {
```
Comment (michalszorad) [inline src/yoroi-wallets/storage.test.ts]: Maybe `getAllKeys, setItem, getItem, removeItem, clear with prefix`  so the text isn't the same as above

File: src/yoroi-wallets/cardano/utxoManager.ts (hunk 19 above)
Comment (wolverineks) [reply src/yoroi-wallets/cardano/utxoManager.ts:107]: thats the plan, just havent gotten around to it
https://github.com/colinhacks/zod

Comment (wolverineks) [reply src/yoroi-wallets/cardano/utxoManager.ts:107]: it wasnt checked before either, i figure its at least a step in the right direction

File: storybook/mocks/wallet.ts (hunk 21 above)
Comment (wolverineks) [reply storybook/mocks/wallet.ts]: leftovers from previous test strategy

**Review by michalszorad — APPROVED** [review]

**Pull request merged**